optional_plugins/result_upload/avocado_result_upload/__init__.py

```diff
@@ -0,0 +1,113 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Virtuozzo Inc. 2017
+# Authors: Dmitry Monakhov <dmonakhov@openvz.org>
+
+"""
+Avocado Plugin to propagate Job results to remote host
+"""
+
+import os
+import sys
+import time
+
+import resultsdb_api
+from six import iteritems
+
+from avocado.core.plugin_interfaces import CLI, Result
+
+from avocado.core.settings import settings
+from avocado.utils import process
+
+
+class ResultUpload(Result):
+
+    """
+    ResultsUpload output class
+    """
+
+    name = 'result_upload'
+    description = 'ResultUpload result support'
+
+    def render(self, result, job):
+        """
+        Upload result, which corresponds to one test from
+        the Avocado Job
+
+        if job.status == "RUNNING":
+            return  # Don't create results on unfinished jobs
+
+        """
+        self.upload_url = None
+        if getattr(job.args, 'result_upload_url', None) is not None:
+            self.upload_url = job.args.result_upload_url
+
+        self.upload_cmd = None
+        if getattr(job.args, 'result_upload_cmd', None) is not None:
+            self.upload_cmd = job.args.result_upload_cmd
+
+        if self.upload_url is None:
+            return
+        if self.upload_url is None:
+            return
+        ret = process.run("%s %s %s" % (self.upload_cmd, job.logdir,
+                                        self.upload_url))
+        if ret.exit_status:
+            job.log.error("ResultUploader failed msg=%s" % result.stderr)
+
+
+class ResultUploadCLI(CLI):
+
+    """
+    Propagate Job results to Resultsdb
+    """
+
+    name = 'resultsdb'
+    description = "Resultsdb options for 'run' subcommand"
+
+    def configure(self, parser):
+        run_subcommand_parser = parser.subcommands.choices.get('run', None)
+        if run_subcommand_parser is None:
+            return
+
+        msg = 'result-upload options'
+        parser = run_subcommand_parser.add_argument_group(msg)
+        parser.add_argument('--result-upload-url',
+                            dest='result_upload_url', default=None,
+                            help='Specify the result upload url')
+
+        def_ssh = 'ssh -o stricthostkeychecking=no -o userknownhostsfile=/dev/null'
+        def_ssh = def_ssh + ' -o batchmode=yes -o passwordauthentication=no'
+        def_upload_cmd = 'rsync -arz -e \'%s \'' % def_ssh
```

Comment:
Ok, this improves the situation a lot. Just a last adjustment, let's handle the `rsync` command availability. We have an utils for that:

```python
from avocado.utils import path as utils_path
...

    try:
        rsync_bin = utils_path.find_command('rsync')
    except CmdNotFoundError:
        #handle missing command

...
```

Reply:
reasonable. Will do